src/Stratis.Bitcoin/Consensus/ConsensusManger.cs

```diff
@@ -319,8 +322,121 @@ private void NotifyChainedHeaderBehaviorsOnReorg()
             this.logger.LogTrace("(-)");
         }
 
-        private async Task<bool> FullyValidateAndReorgAsyncLocked()
+        private async Task<bool> FullyValidateAndReorgLockedAsync(ChainedHeaderBlock chainedHeaderBlock)
         {
+            this.logger.LogTrace("({0}:'{1}')", nameof(chainedHeaderBlock), chainedHeaderBlock);
+
+            ChainedHeader oldTip = this.Tip;
+            ChainedHeader newTip = chainedHeaderBlock.ChainedHeader;
+
+
+            ChainedHeader fork = oldTip.FindFork(newTip);
+            Queue<ConsensusStoreTransitionState> consensusStoreTransitionQueue = null;
+
+
+            if (fork == newTip)
+            {
+                // New header is behind current tip this is a bug
+                throw new Exception(); // TODO: what do we do.
+            }
+
+            // If the new block is not on the current chain as our current consensus tip
+            // then rewind consensus tip to the common fork (or earlier because rewind might jump a few blocks back)
+            if (fork != oldTip)
+            {
+                ChainedHeader current = oldTip;
+                consensusStoreTransitionQueue = new Queue<ConsensusStoreTransitionState>();
+
+                while (fork.Height < current.Height)
+                {
+                    ConsensusStoreTransitionState transitionState = await this.consensusRules.RewindAsync().ConfigureAwait(false);
+                    current = this.chainedHeaderTree.GetChainedHeaderBlock(transitionState.BlockHash).ChainedHeader;
+                    consensusStoreTransitionQueue.Enqueue(transitionState);
+                }
+
+                fork = current;
+            }
+
+            ChainedHeader[] newChain = newTip.ToChainedHeaderArray(fork);
+
+            ChainedHeader lastValidatedHeader = null;
+            ValidationContext validationContext = null;
+            List<int> badPeers = null;
+
+            foreach (ChainedHeader nextChainedHeader in newChain)
+            {
+                if (nextChainedHeader.BlockValidationState != ValidationState.PartiallyValidated)
```

Comment:
It can be fully validated. I'm not sure if we need such a check. But if you want the sanity check, just log it and throw